fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchange.kt

```diff
@@ -0,0 +1,59 @@
+package com.github.dedis.popstellar.model.network.method.message.data.federation
+
+import com.github.dedis.popstellar.model.network.method.message.data.Action
+import com.github.dedis.popstellar.model.network.method.message.data.Data
+import com.github.dedis.popstellar.model.network.method.message.data.Objects
+import com.github.dedis.popstellar.utility.MessageValidator
+import com.google.gson.annotations.SerializedName
+
+/** Token exchange to be broadcast in the LAO */
+class TokensExchange
+/**
+ * Constructor for a data TokenExchange
+ *
+ * @param laoId ID of the remote LAO
+ * @param rollCallId ID of the rollCall of the remote LAO
+ * @param tokens array of tokens contained in the rollCall
+ * @param timestamp timestamp of the message
+ */
+(
+    @SerializedName("lao_id") val laoId: String,
+    @SerializedName("roll_call_id") val rollCallId: String,
+    val tokens: Array<String>,
+    val timestamp: Long
+) : Data {
+
+  init {
+    MessageValidator.verify().isBase64(rollCallId, "rollCallId")
```

Comment:
isBase64 (as every other functions in MessageValidator) returns the builder so you can write : 

`MessageValidator.verify().isBase64(rollCallId, "rollCallId").validPastTimes(timestamp)`

Reply:
given the documentation of the protocol :

<details>
<summary><code>consensus</code></summary>

```
// ../protocol/query/method/message/data/dataFederationTokensExchange.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationTokensExchange.json",
  "description": "Sent by an organizer client to its server, to broadcast the Pop tokens",
  "type": "object",
  "properties": {
    "object": {
      "const": "federation"
    },
    "action": {
      "const": "tokens_exchange"
    },
    "lao_id": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "Hash : HashLen(organizer, creation, name)"
    },
    "roll_call_id": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "last roll call id"
    },
    "tokens": {
      "description": "[Array[Base64String]] list of Pop tokens",
      "type": "array",
      "uniqueItems": true,
      "items": {
        "type": "string",
        "contentEncoding": "base64"
      },
      "$comment": "List must be sorted according to byte encoding: -,0...9,A...Z,_,a...z"
    },
    "timestamp": {
      "type": "integer",
      "description": "[Timestamp] of the tokens' exchange",
      "minimum": 0
    }
  },
  "additionalProperties": false,
  "required": [
    "object",
    "action",
    "lao_id",
    "roll_call_id",
    "tokens",
    "timestamp"
  ]
}
```


</details>

- It looks like that laoId should be checked for `notEmptybase64`.
- I am not sure if an empty rollCallId makes sense, if not then switch to `notEmptybase64 `as well.
- Tokens look like they individually need to be checked as well (they should be base64, maybe we wan't them not empty as well?). Also they should be checked for uniqueness. Is a null or empty array acceptable? This maybe needs to be defined.